### ISSUE TITLE
Unexpected deprecation warning in Rails 6:  'Initialization autoloaded the constant Example'

### ISSUE THREAD

**Issue by ansonhoyt** [rails/rails#36363]
### Steps to reproduce

1. Create a new app with Rails 6.0.0.rc1 (or tip of `6-0-stable` branch)
2. The app initializes fine. Confirm with `rails test`.
3. Define a simple ActiveModel::Type
```
# app/types/example.rb
class Example < ActiveModel::Type::String
end

# config/initializers/types.rb
ActiveRecord::Type.register(:example, Example)
```
4. Now the app initializes with deprecation warnings. Confirm with `rails test`.

### Expected behavior
No Rails deprecation warnings. This is a bare app, using a simple Rails feature as documented.

### Actual behavior
```
$ rails test
DEPRECATION WARNING: Initialization autoloaded the constant Example.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload Example, for example,
the expected changes won't be reflected in that stale Class object.

This autoloaded constant has been unloaded.

Please, check the "Autoloading and Reloading Constants" guide for solutions.
 (called from <top (required)> at /Users/ahoyt/code/tiu/testApp/config/environment.rb:5)

```

### System configuration
**Rails version**: 6.0.0.rc1 (or with tip of `6-0-stable` branch)

**Ruby version**: 2.5.5

**Comment by ansonhoyt** [comment]
Given this warning happens with simple ActiveModel types, I wonder if my authlogic issue (https://github.com/binarylogic/authlogic/issues/672) is really a Rails issue.

**Comment by louim** [comment]
Hi!

My understanding of this problem is that  when the application encounter `ActiveRecord::Type.register(:example, Example)` in your initializer, this is the first time `Example` is referenced, so Rails will try to autoload it. As the deprecation warning message says, Autoloading during initialization is not good because it will not reload correctly.

Wrapping your initializer with `ActiveSupport.on_load(:active_record)` will wait until ActiveRecord has finished loading. By now, your file `example.rb` would have been autoloaded as part of the `app` folder being loaded. When the content of the initializer is executed, it  will not be the first reference to `Example`, so rails won't have to try to autoload it. This also has the advantage of preventing an early load of ActiveRecord.

In your `config/initializers/types.rb`:
```
ActiveSupport.on_load(:active_record) do
  ActiveRecord::Type.register(:example, Example)
end
```

@fxn or somebody more knowledgeable in Rails Autoloading could probably give you a more authoritative answer.

**Comment by fxn** [comment]
If you edit _app/types/example.rb_, the registered class object will become stale, won't reflect the changes. One way to do this cleanly is to move the _types_ directory to _lib_, and perform a regular `require` in the initializer.

**Comment by glyoko** [comment]
I just ran into this same issue when trying to monkey patch a gem in an initializer by including a custom module which I have in lib/. After doing some research it looks like the decision was made to disable autoloading in initializers in the future. Conversation discussing this is [here](https://github.com/rails/rails/pull/35745) and  documented [here](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#autoloading-and-initializers).

To clarify, is the expectation that we explicitly require deps in our initializers? Possible alternative is to use [`require_dependency`](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#require-dependency).

**Comment by fxn** [comment]
Not quite. Let me explain.

`require_dependency` should not be used, that helper is related to autoloading in classic mode, and it keeps track of reloadable constants. Applications running in `zeitwerk` mode should not use `require_dependency`.

What you want, is to load the type, register it, and done. You do not want the class object stored in the constant to become stale if you edit the file and reload. You want the class object to be valid for the entire lifetime of the process regardless of reloads.

So, you have _lib/types/example.rb_, and in the initializer you

```ruby
require 'types/example'
```

which is going to work because _lib_ belongs to `$LOAD_PATH`.

**Comment by glyoko** [comment]
Thanks @fxn for helping to sort that out. Might suggest adding something about requiring deps in initializers in the [autoloading](https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading) section of the upgrade guide.

For @ansonhoyt , I think he just needs `require 'example'` since "app/types/" should already be autoloaded.

**Comment by fxn** [comment]
> For @ansonhoyt , I think he just needs require 'example' since files in "app/types/" should already be autoloaded.

Not really. It would be confusing and a potential issue to have a directory in `autoload_paths` for files that do not have to be autoload nor reloaded.

As I said above, better move that to _lib_.

**Comment by glyoko** [comment]
You're correct. Mistakenly thought that app/types/ was already in $LOAD_PATH in a fresh rails app to work with `ActiveRecord::Type`.

**Comment by ansonhoyt** [comment]
Ah, of course, that makes sense! Thanks @fxn for the explanation!

I forgot that any `app/*` paths are autoloaded! So, I introduced this autoloaded path into the types initializer...and fully deserved the warning.
```
$ spring stop
$ rails r 'puts ActiveSupport::Dependencies.autoload_paths'
...
/Users/ahoyt/code/testApp/app/types
```

The deprecation warning about autoloaded constants went away when I made moved my types to `lib` and simply required them directly:
```
# lib/types/example.rb
class Example < ActiveModel::Type::String
end

# config/initializers/types.rb
Dir[Rails.root.join('lib', 'types', '**', '*.rb')].each { |f| require f }
ActiveRecord::Type.register(:example, Example)
```
Now, all my types work, my tests pass, and neither `app/types` or `lib/types` are in my autoload_paths.

Happy dance! 😀 🎉 🕺 🥂

**Comment by fxn** [comment]
Fantastic 😀.

**Comment by viktorMarkevich** [comment]
In my case the deprecation warning occurs for  **require "action_text/engine"** row in application.rb

**Comment by jackkinsella** [comment]
Are there up-to-date docs about how to get Zeitwerk to play well with `ActiveRecord::Type`. All I found so far was this: https://www.rubydoc.info/gems/activerecord/ActiveRecord/Type

**Comment by ansonhoyt** [comment]
There is a Rails Guide, [Autoloading and Reloading Constants (Zeitwerk Mode)](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html). Check out section
[6.1 Reloading and Stale Objects](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#reloading-and-stale-objects).

The docs for the [attribute](https://api.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html#method-i-attribute) method are pretty good for how to work with types, but it doesn't warn about the pitfalls of how you load your custom types. That would make a good addition!

To fix this exception, just make sure you don't autoload your custom types, like I did, but put them in `/lib`. See [comment above](https://github.com/rails/rails/issues/36363#issuecomment-499992270) for how I load my custom types.

btw, the above advice from @fxn is solid. He wrote zeitwerk.

**Comment by jackkinsella** [comment]
@ansonhoyt thanks for a v helpful answer. 

I'd gladly volunteer to update the docs, at least if I can get my head around one confusion: why is it required that `ActiveRecord::Type` subclasses have to only be loaded once (and therefore, post `Zeitwerk`, need to be relegated to the `lib` folder)?  

Most of the Ruby action in a Rails app happens in the `app` folder and it's all happily autoloaded, making for a wonderful development experience where you can carry out continuous experiments in development without needing to restart the server. Why do `ActiveRecord::Type` subclasses need special treatment? Is it performance related, due to load order constraints, or "that's just the way it is"?

**Comment by fxn** [comment]
> Most of the Ruby action in a Rails app happens in the app folder and it's all happily autoloaded, making for a wonderful development experience where you can carry out continuous experiments in development without needing to restart the server. Why do ActiveModel::Type subclasses need special treatment? Is it performance related, due to load order constraints, or "that's just the way it is"?

This is not something related to Zeitwerk.

The type is registered in an initializer. No matter whether you use `classic` or `zeitwerk` mode, if you edit your custom type and reload, changes won't take effect because initializers run only when the application boots.

So before Rails 6, you could _think_ the edit had effect, but it did not. Advising against this is the point of the warning in Rails 6.

I believe it should be possible to get it working if you registered at the bottom of the file that implements the custom type (because I believe the register would override the entry with the same key), and configured a callback to force loading that constant. But I don't have time right now to test that idea.

**Comment by jackkinsella** [comment]
@fxn Oops, I think there was a bit of a miscommunication above – I didn't think this had anything to do with Zeitwerk – I'm aware it's a Rails thing. I was hoping for some elaboration on why the customer types had to be registered in an initializer.

**Comment by fxn** [comment]
@jackkinsella could you please try this?

1) Define `CurrencyType` as you did before, without the line that registers.
2) In an initializer register in a `to_prepare` block:

    ```ruby
    Rails.application.config.to_prepare do
      ActiveRecord::Type.register(:currency, CurrencyType)
    end
     ```

If `register` overwrites the previous `:currency` entry as I believe, that should do the trick and the class could be reloadable.

**Comment by jackkinsella** [comment]
Thanks for taking interest. 

**TLDR: The approach was a step in the right direction, but still didn't work due to a singleton object in `ActiveModel::Type` holding registration state across reloads combined with a priority function for choosing between competing type implementations that does not account for order of addition.**

## My Code

```ruby
# config/initializers/types.rb
Rails.application.config.to_prepare do
  ActiveRecord::Type.register(:currency, CurrencyType)
end

# app/types/currency_type.rb
class CurrencyType < ActiveRecord::Type::Decimal
  def cast(value)
    ...
  end
end 
# EOF

# app/models/tutor.rb
class Tutor < ActiveRecord::Base
  attribute :price_per_hour, :currency
end
```

## Debugging the type registration system in ActiveRecord

Then in the Rails console (without spring - `DISABLE_SPRING=1 rails console`), when I type `reload!` after modifying the `CurrencyType`, the latest changes were not loaded.

With a debugger, I traced the issue to [active_record/type/adapter_specific_registry.rb](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/type/adapter_specific_registry.rb). After reloading the code in the console, the `registrations` array [see this line](https://github.com/rails/rails/blob/056c1bb070ba2b4e6464409bd59fbaac345891c1/activerecord/lib/active_record/type/adapter_specific_registry.rb#L19), which is first initialized in the superclass `ActiveModel::Type::Registry`[(view line)](https://github.com/rails/rails/blob/056c1bb070ba2b4e6464409bd59fbaac345891c1/activemodel/lib/active_model/type/registry.rb#L8) , increases in size after each reload (oddly, by 2, twice the number of calls to `register` in my `config.to_prepare` block. This continued holding of state across reloads smelled of singleton pattern, and sure enough the whole Registry itself is defined as a class instance variable on `ActiveModel::Type` [(view line)](https://github.com/rails/rails/blob/056c1bb070ba2b4e6464409bd59fbaac345891c1/activemodel/lib/active_model/type.rb#L22)

When there are multiple types registered to a symbol (e.g. to `:currency`) it chooses between them based on a priority given by a `<=>` method that does not take into account the order in which the type was registered. When I manually added on a field to the type registration based on time of registration and modified the priority function to use this, the code reloading worked fine.

So, if I'm not mistaken, it looks like we have two options to get type reloading working:
* ditch the singleton (assuming it won't hurt performance too badly)
* rewrite the priority function to account for later registrations (downside: would be a memory leak)

**Comment by jasonperrone** [comment]
So, here's something related and I am not sure what to do about it. I have an initializer (for Devise, actually) where I have some custom code. The code in the initializer calls an ActiveRecord class because I configure Devise based on a value I have stored in the database. Let's say my class is called Dog.

In the initializer, I simply say Dog.find(1).name.

All I am doing is referencing this AR class because I need the database value for my initializer config and this sparks the DEPRECATION warning that initialization autoloaded the reloadable constant Dog. Of course it's reloadable it's in /app/models. 

What do I do? Does this mean I am basically not allowed to refer to ActiveRecord classes in my initializers at all?

**Comment by fxn** [comment]
Hey @jasonperrone, could you please read sections 6.1 and 6.2 in the [edge autoloading guide](https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html#autoloading-when-the-application-boots), and see if that solves your use case?

**Comment by jasonperrone** [comment]
Just noticed 6.2. Will try!

**Comment by jasonperrone** [comment]
"Rails.application.reloader.to_prepare do" seems to have worked for me @fxn , thank you.

**Comment by fxn** [comment]
Fantastic! 👍
